### DISCUSSION TITLE
fire rain // SPL_MassDeath // SPL_IceWave

### DISCUSSION THREAD

**Discussion by Kor-Angar** [dosinabox/g2nr_unofficial_update#453] [General]
workshop Spell FireRain Fix 
https://steamcommunity.com/sharedfiles/filedetails/?id=2825215672

**Comment by Kor-Angar** [comment]
Бонусы и штрафы от огненного дождя

```D
//----- Feuer -----	
	
	if (spellType 	== SPL_ChargeFireball)
	|| (spellType  	== SPL_InstantFireball)		
	|| (spellType  	== SPL_Firerain)		
	|| (spellType 	== SPL_Firebolt)		
	|| (spellType 	== SPL_Firestorm)
	|| (spellType   == SPL_Pyrokinesis)
	|| (spellType	== SPL_Deathbolt)
	|| (spellType 	== SPL_Deathball)
	{
		if (C_NpcIsDown(self))
		|| (C_BodyStateContains(self,BS_SWIM)) 	
		|| (C_BodyStateContains(self,BS_DIVE))
		{
			return COLL_DONOTHING;
		};
		
		// feuer monster bekommen halben schaden, und kein opfer zs
		if (self.guild == GIL_FIREGOLEM	) 		
		|| (self.aivar[AIV_MM_REAL_ID]	== 	ID_FIREWARAN)
		|| (self.guild == GIL_GARGOYLE)
		|| (self.aivar[AIV_MM_REAL_ID] == ID_DRAGON_FIRE) 
		{
			return COLL_APPLYHALVEDAMAGE;
		}
		
		// grosse eis monster kriegen den doppelten schaden, brennen aber nicht
		if (self.guild == GIL_ICEGOLEM)
		|| (self.aivar[AIV_MM_REAL_ID] == ID_DRAGON_ICE) 
		|| (self.aivar[AIV_MM_REAL_ID] == ID_ICEWOLF) 
		{
			return COLL_APPLYDOUBLEDAMAGE;
		};
					
		// alle grosse monster bekommen nur schaden, kein opfer zs
		if (self.guild == GIL_STONEGOLEM)
		|| (self.guild == GIL_SUMMONED_GOLEM)
		|| (self.guild == GIL_DEMON)
		|| (self.guild == GIL_SUMMONED_DEMON)
		|| (self.guild == GIL_TROLL)
		|| (self.guild == GIL_DRAGON) //andere
		{	
			return COLL_APPLYDAMAGE;
		};
		
		return COLL_DOEVERYTHING;
	};
```

**Reply by Kor-Angar** [reply]
бонусы и штрафы от огня
```D
func void B_MagicFireHurtNpc2022 (var C_NPC attacker, var C_NPC slf, var int damage)
{

	// feuer monster bekommen halben schaden, und kein opfer zs
	if (self.guild == GIL_FIREGOLEM	) 		
	|| (self.aivar[AIV_MM_REAL_ID]	== 	ID_FIREWARAN)
	|| (self.guild == GIL_GARGOYLE)
	|| (self.aivar[AIV_MM_REAL_ID] == ID_DRAGON_FIRE) 
	{
		damage = damage / 2;
		// без эффекта горения
	}
	// grosse eis monster kriegen den doppelten schaden, brennen aber nicht
	else if (self.guild == GIL_ICEGOLEM)
	|| (self.aivar[AIV_MM_REAL_ID] == ID_DRAGON_ICE) 
	|| (self.aivar[AIV_MM_REAL_ID] == ID_ICEWOLF) 
	{
		damage = damage * 2;
		if 	(self.aivar[AIV_MM_REAL_ID] == ID_ICEWOLF)
		{
			// ледяной волк наверное должен гореть?
			Wld_PlayEffect ( "VOB_MAGICBURN", slf, slf, 0, 0, 0, FALSE );
		};
	}		
	// alle grosse monster bekommen nur schaden, kein opfer zs
	else if (self.guild == GIL_STONEGOLEM)
	|| (self.guild == GIL_SUMMONED_GOLEM)
	|| (self.guild == GIL_DEMON)
	|| (self.guild == GIL_SUMMONED_DEMON)
	|| (self.guild == GIL_TROLL)
	|| (self.guild == GIL_DRAGON) //andere
	{
		// нет эффекта  горения
	}
	else
	{	
		// эффект горения
		Wld_PlayEffect ( "VOB_MAGICBURN", slf, slf, 0, 0, 0, FALSE );
	};
	
	var int effectiveDamage;
	
	effectiveDamage = damage - slf.protection[PROT_MAGIC];
	
	if (effectiveDamage < 0)
	{
		effectiveDamage = 0;
	};

	//Npc_ChangeAttribute (slf,	ATR_HITPOINTS, -effectiveDamage);
	B_MagicHurtNPC (attacker,slf,effectiveDamage);

	//B_StopMagicHurtNpc2022();
};

```

**Reply by Kor-Angar** [reply]
ледяные големы мрут с одного дождя
![20220624194532_1](https://user-images.githubusercontent.com/49514764/175606678-9c4cb7a0-92b3-4dbe-ae57-82d8fde26789.jpg)
![20220624194543_1](https://user-images.githubusercontent.com/49514764/175606702-29f7214b-9c00-4523-940c-bd838070282d.jpg)

**Comment by Kor-Angar** [comment]
в **Spell_Cast_Firerain** запускается ZS_Firerain, определяется радиус. тут же можно запустить альтернативный сценарий если дождь кастует ищущий.

```D
func void Spell_Cast_Firerain()
{
	PrintScreen			("Spell_Cast_Firerain",-1,35,FONT_ScreenSmall,5);
	
	// если кастер ГГ
	//	if (Npc_IsPlayer(self))
	var C_NPC her; 	her = Hlp_GetNpc(PC_Hero); 
	
	if  (Hlp_GetInstanceID(self) == Hlp_GetInstanceID(her))
	{	
		AI_SetNpcsToState	(self, ZS_Firerain, 1000);
	}
	// проверка на гильдию ищущих
	else
	{
	
	};	
	Spell_Cast_Firerain_old ();	// отнимает ману у героя
};

```

**Reply by Kor-Angar** [reply]
ввиду того что поменял 	`targetCollectAlgo		= TARGET_COLLECT_CASTER;` стало  непонятно откуда **лучше**  запускать 		`AI_SetNpcsToState	(self, ZS_Firerain, 1000);` есть два варианта

- [ ] в Spell_Cast_Firerain
- [ ] или в C_CanNpcCollideWithSpell 

пока оставлю запуска через `Spell_Cast_Firerain`

**Reply by Kor-Angar** [reply]
через `C_CanNpcCollideWithSpell ` оно работает всё равное некорректно :-)

**Reply by Kor-Angar** [reply]
это искусственная проблема про то что ищущих не должен убивать дождь от ищущих. пусть убивает. мне кажется это не та проблема которая требует решения.

**Comment by Kor-Angar** [comment]
видос 

https://www.youtube.com/watch?v=QHSSfCAxIW8

**Comment by Kor-Angar** [comment]
Радиус какой то нужно придумать 1000 это мало, а 10000 много :-)

**Reply by Kor-Angar** [reply]
пока 2000 оптимально. можно еще чутка подумать :-)

**Comment by Kor-Angar** [comment]
workshop Spell FireRain Fix 
https://steamcommunity.com/sharedfiles/filedetails/?id=2825215672

**Comment by Kor-Angar** [comment]
файлы

**Reply by Kor-Angar** [reply]
[KA_Spell_FireRain.zip](https://github.com/dosinabox/g2nr_unofficial_update/files/8980254/KA_Spell_FireRain.zip)

**Comment by Kor-Angar** [comment]
наверное НПС должны  адекватно реагировать на героя после того он кастанул огненный дождь :-)
```D
// Функция завершения состояния
func void ZS_Firerain_End()
{
	B_Attack (self, hero, AR_KILL,0); 
};
```

**Comment by Kor-Angar** [comment]
надо будет в игре понаблюдать как работает ледяная волна и волна смерти с фиксом на уровне движка https://worldofplayers.ru/threads/41521/page-5#post-1075714 . сравнивая с тем как работает огненный дождь с фиксом на скриптах. попытаться понять чего не хватает дождю что бы он обходился без скрпитов.

**Comment by Kor-Angar** [comment]
из плюсов можно отметить что движок выведет жертву из цикла zs_ при окончании действия заклинания. например у огненного дождя это 6 с хвостиком секунд. при этом этот плюс так себе плюс, так как выводит он его жёстко без завершения , то есть без запуска всего что то прописано в скриптах. таким образом в отдельно взятых ситуациях это может быть и минусом.  так например жертва может не с агрится на кастера. вообщем надо играть и смотреть. смотреть и играть.

**Reply by Kor-Angar** [reply]
> из плюсов

фикса SpellFix,dll

**Comment by Kor-Angar** [comment]
как вариант можно упростить код для огненного дождя за счёт фикса SpellFix,dll 

1. в `C_CanNpcCollideWithSpell `ловить людей 
2.  в `C_CanNpcCollideWithSpell `отправлять людей в состояние `ZS_Firerain` отдавая  движку 		`return COLL_DONOTHING;`
3.  `ZS_Firerain`  использовать только для людей: наносить им там ущерб, делать анимацию горения и прочие радости

**Reply by Kor-Angar** [reply]
так и правда работает. кода меньше. НО огненный дождь разит хуже :-)

**Reply by Kor-Angar** [reply]
на скриптах было бодрее так как попадало по всем :-) а на движке "промахивается" если нпс сильно занят своим делом :-) https://www.youtube.com/watch?v=f3PjOzap0yw

**Reply by Kor-Angar** [reply]
ну и ладно. лучшее это враг  хорошему :-)
[UNION_SPELLFIX+script.zip](https://github.com/dosinabox/g2nr_unofficial_update/files/9198047/UNION_SPELLFIX%2Bscript.zip)
.

**Comment by Kor-Angar** [comment]
@dosinabox  этот кусок в Готики 2 Ночь Ворона работать не будет / не должен.  это атавизм готики 2 классик.

**ZS_MagicFreeze** 
```D
if((self.guild == GIL_FIREGOLEM) || (self.aivar[AIV_MM_REAL_ID] == ID_FIREWARAN) || (self.aivar[AIV_MM_REAL_ID] == ID_DRAGON_FIRE))
			{
				B_MagicHurtNpc(other,SPL_FREEZE_DAMAGE * 2);
				return LOOP_CONTINUE;
			};
			if((self.guild == GIL_ICEGOLEM) || (self.aivar[AIV_MM_REAL_ID] == ID_DRAGON_ICE))
			{
				B_MagicHurtNpc(other,SPL_FREEZE_DAMAGE / 2);
				return LOOP_CONTINUE;
			};
```

**Reply by Kor-Angar** [reply]
мне кажется можно добавить отдельную функцию проверку на то кого нельзя заморозить взяв список из C_CanNpcCollideWithSpell.d

а в ZS_MagicFreeze делать на всякий случай проверку с выходом из цикла для тех существ которых нельзя заморозить. т.е. я так делал в рамках фикса "ледяной волны" на уровне скриптов. а потом всё это убрал так как появился файл dll.

**Reply by Kor-Angar** [reply]
> Тот же вопрос про ледяного волка и SPL_FREEZE_DAMAGE / 2. Тоже забыли добавить к ледяному голему и ледяному дракону?)

тупо сделали запрет выше уровнем в C_CanNpcCollideWithSpell , а файл ZS_MagicFreeze  не поправили так как он перестал для них работать. было бы правильно с их стороны закомментировать кусок который перестал работать и написать краткий комментарий что это для Ночь Ворона уже не актуально сославшись на файл C_CanNpcCollideWithSpell.

**Reply by Kor-Angar** [reply]
> а в ZS_MagicFreeze делать на всякий случай проверку с выходом из цикла для тех существ которых нельзя заморозить

для лука писал проверку
```D
func int MagicFreezeGo(var c_npc npc)
{
	if (C_NpcIsDown(npc))
	|| (C_BodyStateContains(npc,BS_SWIM)) 	
	|| (C_BodyStateContains(npc,BS_DIVE))
	{
		return false;
	};
	if (npc.guild == GIL_ICEGOLEM)
	|| (npc.aivar[AIV_MM_REAL_ID] == ID_DRAGON_ICE) // избыточно
	|| (npc.aivar[AIV_MM_REAL_ID] == ID_ICEWOLF)
	{
		return false;
	};
	if (npc.guild == GIL_STONEGOLEM)
	|| (npc.guild == GIL_SUMMONED_GOLEM)
	|| (npc.guild == GIL_DEMON)
	|| (npc.guild == GIL_SUMMONED_DEMON)
	|| (npc.guild == GIL_TROLL)
	|| (npc.guild == GIL_DRAGON) //andere
	{
		return false;
	};
	
	if (npc.guild == GIL_FIREGOLEM)
	|| (npc.aivar[AIV_MM_REAL_ID]	== 	ID_FIREWARAN)
	|| (npc.guild == GIL_GARGOYLE)
	|| (npc.aivar[AIV_MM_REAL_ID] == ID_DRAGON_FIRE) // избытычно
	{
		return false;
	};
	return true;
};
```

**Reply by Kor-Angar** [reply]
тут они разделены на три группы что бы было проще сверить с источником. соответственно если делать проверку то лучше по твоей схеме построчно для каждого

`if (npc.aivar[AIV_MM_REAL_ID] == ID_ICEWOLF)	{ return false;	};`

**Comment by Kor-Angar** [comment]
Spell Mass Death Fix
*не действует на нежить, в том числе дракона нежить;
*драконам наносит половину ущерб;

**Reply by Kor-Angar** [reply]
у меня в  мини моде стоит просто значение 1000.

**Reply by Kor-Angar** [reply]
тут короче всё просто - волна смерти убивает. вот и вся история.
на нежить она не действует.
дракону наносит половину ущерба. видимо речь идёт про огненного дракона в 6ой главе.
что с драконом нежитью в части идеологии не понятно. то ли не действует то ли половина ущерба.

как то так.

**Reply by Kor-Angar** [reply]
в связи с чем я думаю можно сделать вот так :-)

```D
    if (spellType == SPL_MassDeath)
    {
        if (C_NpcIsUndead(self))               { return COLL_DONOTHING;	}; // не действует на нежить
        
        if (self.guild == GIL_DRAGON)      { return COLL_APPLYHALVEDAMAGE;	}; // тут наверное речь идёт про дракона нежить?? или про огненного дракона? дракон нежить проходит проверку по C_NpcIsUndead ?
        
        self.attribute[ATR_HITPOINTS] = 13;

        return COLL_DOEVERYTHING;
    };
```

**Reply by dosinabox** [reply]
> дракон нежить проходит проверку по C_NpcIsUndead ?

да

**Reply by dosinabox** [reply]
```cpp
// ---- Todeswelle ----
	
	if (spellType == SPL_MassDeath)
	{
		if (!C_NpcIsUndead(self))	
		{
			if (self.guild == GIL_DRAGON)
			{
				return COLL_APPLYHALVEDAMAGE;
			};
			return COLL_DOEVERYTHING;
		};
		return COLL_DONOTHING;
	};
```
Так понятнее:
```cpp
if(spellType == SPL_MassDeath)
{
	if(C_NpcIsUndead(self))
	{
		return COLL_DONOTHING;
	};
	if(self.guild == GIL_DRAGON)
	{
		return COLL_APPLYHALVEDAMAGE;
	};
	return COLL_DOEVERYTHING;
};
```

**Reply by dosinabox** [reply]
Всё ещё не понимаю, что именно в обновлении режет урон волны с двойного до обычного. Если она в оригинале всегда наносит 1000 урона, то пусть и в обновлении всегда наносит 1000 урона. Можно даже "официально" признать эту фичу и прямо указать 1000 урона в описаниях свитков и руны. Но нужно знать, от чего это зависит.

**Reply by Kor-Angar** [reply]
волна в игре наносила двойной урон потому что частицы попадали по врагу дважды. ищи какой dll файл или patch это исправляет в твоей сборке.

можно прямо указать 1000 урона в неоф. обновление. я бы в описание вообще не указывал урон, а написал что она всех убивает. в этой связи надо проверить так же главу ищущих.

там дело в том что она просто должна всех убивать. кроме дракона. дракон там получается один - красный.

**Reply by dosinabox** [reply]
> там дело в том что она просто должна всех убивать. кроме дракона. 

И кроме нежити.

**Reply by Kor-Angar** [reply]
кроме нежите это понятно. я про красного дракона на острове которого она не должна убить с одного каста ;-)

то есть или убивает или недействует. исключение только дракон. он концепцию ломает :-)

**Reply by dosinabox** [reply]
В чистом оригинале с одним Юнионом тоже 500 урон.

**Reply by Kor-Angar** [reply]
> В чистом оригинале с одним Юнионом тоже 500 урон.

а нежить убивает?

**Reply by dosinabox** [reply]
В оригинале с юнионом убивает нежить и наносит 500 урона.
В обновлении с юнионом не убивает нежить и наносит 500 урона.

**Reply by Kor-Angar** [reply]
Тогда надо просто ставить тысячу ущерба и не заморачиваться
